etcdserver/server.go

```diff
@@ -632,6 +628,21 @@ func (s *EtcdServer) Lead() uint64 { return atomic.LoadUint64(&s.r.lead) }
 
 func (s *EtcdServer) Leader() types.ID { return types.ID(s.Lead()) }
 
+// RequestTimeout returns an appropriate timeout to wait for request to be committed.
+func (s *EtcdServer) RequestTimeout() time.Duration {
```

Comment:
After going through this again, I think it is etcdserver.Do() 's responsibility to end itself early if it finds out that there is no chance to commit this. The ctx parameter in etcdhttp side is used for early termination or deadline set. So it is fine to use max timeout in etcdhttp side. Sound good?